napari/layers/utils/text_manager.py

```diff
@@ -55,95 +74,97 @@ class TextManager(EventedModel):
-        self._set_text(text, n_text, properties=properties)
 
-    def _set_text(
-        self,
-        text: Optional[str],
-        n_text: int,
-        properties: Optional[dict] = None,
-    ):
-        if properties is None:
-            properties = {}
-        if text is None:
-            text = np.empty(0)
-        if len(properties) == 0 or len(text) == 0:
-            # set text mode to NONE if no props/text are provided
-            self._mode = TextMode.NONE
-            self._text_format_string = ''
-            self.values = np.empty(0)
+    @property
+    def values(self):
+        _warn_about_deprecated_values_field()
+        n_text = _infer_n_text(self.string, self.properties)
+        return self.string._get_array(self.properties, n_text)
+
+    def __setattr__(self, key, value):
+        if key == 'values':
+            _warn_about_deprecated_values_field()
+            self.string = value
         else:
-            formatted_text, text_mode = format_text_properties(
-                text, n_text, properties
-            )
-            self._text_format_string = text
-            self._mode = text_mode
-            self.values = formatted_text
+            super().__setattr__(key, value)
+        if key == 'properties':
+            self._on_properties_changed()
 
-    def refresh_text(self, properties: dict):
-        """Refresh all of the current text elements using updated properties values
+    def refresh_text(self, properties: Dict[str, np.ndarray]):
+        """Refresh all text elements from the given layer properties.
 
         Parameters
         ----------
-        properties : dict
-            The new properties from the layer
+        properties : Dict[str, np.ndarray]
+            The properties of a layer.
         """
-        self._set_text(
-            self._text_format_string,
-            n_text=len(self.values),
-            properties=properties,
-        )
+        self.properties = properties
 
-    def add(self, properties: dict, n_text: int):
-        """Add a text element using the current format string
+    def add(self, properties: dict, num_to_add: int):
+        """Adds a number of a new text elements.
 
         Parameters
         ----------
         properties : dict
-            The properties to draw the text from
-        n_text : int
-            The number of text elements to add
+            The current properties to draw the text from.
+        num_to_add : int
+            The number of text elements to add.
         """
-        if self._mode in (TextMode.PROPERTY, TextMode.FORMATTED):
-            new_text, _ = format_text_properties(
-                self._text_format_string, n_text=n_text, properties=properties
-            )
-            self.values = np.concatenate((self.values, new_text))
+        warnings.warn(
+            trans._(
+                'TextManager.add is a deprecated method. '
+                'Use TextManager.string._get_array(...) instead.'
```

Comment:
Right now this deprecation warning points to something private, is that because we think from now on this stuff should all be private, or do we want to make it public one day? Same here and below

Reply:
Yeah, this demonstrates the tension in this PR. I'm deprecating some behavior because it's hard to support completely with the general approach here (and it's not easy on `main` either), but not replacing it with public behavior because some things are not settled yet (e.g. what state does `properties` have, does the encoding have access to `properties`, does `properties` have access to all the encodings).